src/plugins/clips-executive/clips/resource-locks.clp

```diff
@@ -34,61 +39,86 @@
 )
 
 (defrule resource-locks-request-lock
+  "Request a mutex for each required resource if none of the respective mutexes
+   is already locked or has a pending request."
   (goal (mode COMMITTED)
-        (acquired-resources $?acq)
-        (required-resources $?req))
-  ; We may have multiple goals requiring the same resource. If that's the case,
-  ; wait until the other goal releases the resources.
+        (id ?goal-id)
+        (acquired-resources)
+        (required-resources $?req&:(> (length$ ?req) 0)))
   (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
               (request ~NONE)))
   (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
               (state LOCKED)))
+  (not (resource-request (resource ?res&:(member$ ?res ?req))))
   =>
-  (foreach ?res (set-diff ?req ?acq)
+  (foreach ?res ?req
     (printout warn "Locking resource " ?res crlf)
     (mutex-try-lock-async (resource-to-mutex ?res))
+		(assert (resource-request (goal ?goal-id) (resource ?res)))
   )
 )
 
-(defrule resource-locks-fast-reject-goal
-  "If a resource is locked by someone else, we have not acquired any resources,
-   and we have no pending requests, then we can directly reject the goal."
+(defrule resource-locks-fast-reject-goal-locked-by-other-agent
+  "If a resource already locked by someone else, we have not acquired any
+   resources, and we have no pending requests, then we can directly reject the
+   goal."
   (wm-fact (key cx identity) (value ?identity))
   ?g <- (goal (mode COMMITTED)
               (id ?goal-id)
               (acquired-resources)
               (required-resources $?req))
+  ; The mutex is locked and there is no unprocessed request, which means that
+  ; it was either locked by someone else or for a different goal.
   (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
-         (state LOCKED) (locked-by ?locker&~?identity))
-  (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
+         (state LOCKED) (request NONE) (locked-by ?locker&~?identity))
+  (not (mutex (name ?n1&:(member$ (mutex-to-resource ?n1) ?req))
               (request ~NONE)))
   =>
   (printout warn "Rejecting goal " ?goal-id ", " (mutex-to-resource ?n)
                  " is already locked by " ?locker crlf)
   (modify ?g (mode FINISHED) (outcome REJECTED))
 )
 
-(defrule resource-locks-lock-acquired
-  ?m <- (mutex (name ?res) (request LOCK) (response ACQUIRED))
+(defrule resource-locks-fast-reject-goal-locked-for-another-goal
+  "If a resource already locked for another goal, we have not acquired any
+   resources, and we have no pending requests, then we can directly reject the
+   goal."
   ?g <- (goal (mode COMMITTED)
+              (id ?goal)
+              (acquired-resources)
+              (required-resources $? ?res $?))
+  (resource-request (resource ?res) (goal ?other-goal&~?goal))
+  (not (mutex (name ?n1&:(eq ?n1 (resource-to-mutex ?res)))
+              (request ~NONE)))
```

Comment:
This should check, if n1 is a member of the required resources, I suppose? Otherwise, we cant reject the goal if any mutex has an open request. However, this would make the condition quite ugly, since you would have to match `required-resources ($?rest-1 ?res $?rest-2)` and check for n1 not being a member of ?rest-1, ?rest-2 and being unequal to ?res